pkg/controller/publishingstrategy/publishingstrategy_controller.go

```diff
@@ -105,78 +109,122 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
-			err := r.defaultIngressHandle(appingress, ingressControllerList, newCertificate)
-			if err != nil {
-				log.Error(err, fmt.Sprintf("failed to handle default ingresscontroller %v", appingress))
-				return reconcile.Result{}, err
+		// Compare the Spec fields that cannot be patched in the desired IngressController and the actual IngressController
+		if !validateStaticSpec(*ingressController, desiredIngressController.Spec) {
+			// "The default" CR also needs a status check as the config isn't always guaranteed to be in the spec
+			if ingressDefinition.Default {
+				if !validateStaticStatus(*ingressController, desiredIngressController.Spec) {
+					// Since the default IngressController CR spec + status does not match the desired IngressController
+					// spec that was generated based on the ApplicationIngress, and the fields checked are immutable,
+					// the actual default IngressController must be deleted
+					r.client.Delete(context.TODO(), ingressController)
+					return reconcile.Result{Requeue: true}, nil
+				}
+			} else {
+				// Since the default IngressController CR spec does not match the desired IngressController
+				// spec that was generated based on the ApplicationIngress, and the fields checked are immutable,
+				// the IngressController must be deleted
+				r.client.Delete(context.TODO(), ingressController)
+				return reconcile.Result{Requeue: true}, nil
 			}
-			continue
-		}
-		err := r.nonDefaultIngressHandle(appingress, ingressControllerList, newCertificate)
-		if err != nil {
-			log.Error(err, fmt.Sprintf("failed to handle non-default ingresscontroller %v", appingress))
 		}
-	}
-
-	ingressList := &operatorv1.IngressControllerList{}
-	listOptions = []client.ListOption{
-		client.InNamespace("openshift-ingress-operator"),
-	}
-	err = r.client.List(context.TODO(), ingressList, listOptions...)
-	if err != nil {
-		log.Error(err, "Cannot get list of ingresscontroller")
-		return reconcile.Result{}, err
-	}
 
-	// since we don't own ingresscontroller object, adding this check ensures that the proper ingresscontrollers have been created by cluster-ingress-operator
-	// if check fails, requeue the reconcile and try to re-create the ingresscontrollers until successful
-	if !r.ensureIngressControllersExist(instance.Spec.ApplicationIngress, ingressList) {
-		reqLogger.Info("IngressController does not match PublishingStrategy. Requeue and try again")
-		return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		// All the remaining fields are mutable and don't require a deletion of the IngresscController
+		// If any of the fields are differet, that field will be patched
+		if valid, field := validatePatchableSpec(*ingressController, desiredIngressController.Spec); !valid {
+			baseToPatch := client.MergeFrom(ingressController.DeepCopy())
+			if ingressDefinition.Default {
+				if valid, field := validatePatchableStatus(*ingressController, desiredIngressController.Spec); !valid {
```

Comment:
Why isn't this function used for both? Is it wise to re-scope `field`?

Reply:
Used for both what?

Reply:
For both "default" and "non-default." I suppose I don't understand the purpose of `validatePatchableSpec`.

Reply:
@lisa As explained in the comments

The default IngressController is special. When its first created, the spec is not filled out.
Instead, the relavent fields are set in the status. For each of the spec checks, the status
will also be checked if the ApplicationIngress references the default IngressController

Reply:
@fahlmant I think I was unclear in phrasing my question. What does `validatePatchableStatus` do? The function has no comments. What are the semantic meanings behind its two return values? What inputs cause certain kinds of outputs? Can it be used for both default and non-default ingress types?

It appears that there's duplication adjacent to this comment and if the `validatePatchableStatus` function can service both default and non-default ingresses that some of the duplication can be removed and thus the readability improved somewhat.